### PR TITLE
[event based orderbook] Handle orderbook creation from State with regards to batch id better

### LINKED ISSUE #754: [event based orderbook] Nick's idea for better modelling getter methods in terms of batch id
On a sort of separate note, unrelated to this PR, I had a nice thought for modelling the getter methods in terms of batch IDs with respect to the `balances.get_balance(0)` not actually returning the balance at batch ID 0. The parameter there can be:
```
Batch {
    Solving,
    Future(std::num::NonZeroU32),
}
impl Batch {
    fn from_id(current_batch_id, batch_id) {
        ensure!(current_batch_id <= batch_id, "batch in the past");
        std::num::NonZeroU32::new(batch_id - current_batch_id)
            .map(Batch::Future)
            .unwrap_or(Batch::Solving)
    }
}
```

I'm still not 100% satisfied with this, but then it would make it impossible to try and get the balance in a past batch and misuse the API.

_Originally posted by @nlordell in https://github.com/gnosis/dex-services/pull/753#issuecomment-617545797_

### PR DESCRIPTION
**Pull request by e00E**
fixes #754

This prevents attempting to create the orderbook with a past batch with
is nonsensical as State does not actually support.
This prevents reading the orderbook for a future batch if that
orderbook would be inconsistent due to a partially applied solution.

I notice that the tests feels needlessly long. In the future I want to refactor the tests into more helper functions and macros to make them smaller and easier to read.

### CODE REVIEW

File: driver/src/orderbook/streamed/state.rs [hunk 1]
```diff
@@ -23,6 +23,18 @@ pub struct State {
     balances: HashMap<(UserId, TokenAddress), Balance>,
     tokens: Tokens,
     last_solution: LastSolution,
+    /// True when we have received some trades or trade reverts but not yet the final solution
+    /// submission that indiciates all trades have been received.
+    solution_partially_received: bool,
```
Comment (fleupold) [inline driver/src/orderbook/streamed/state.rs:28]: Maybe this fits better directly on the LastSolution struct or could be expressed by making last_solution an option that is None when it's in the process of receiving (although I'm not sure that's very explicit).

File: driver/src/orderbook/streamed/state.rs [hunk 2]
```diff
@@ -33,8 +45,31 @@ struct LastSolution {
...
+        &self,
+        batch: Batch,
+    ) -> Result<(
+        impl Iterator<Item = ((UserId, TokenId), U256)> + '_,
+        impl Iterator<Item = ModelOrder> + '_,
+    )> {
+        let batch_id = match batch {
+            Batch::Current => self.last_batch_id,
+            Batch::Future(batch_id) => {
+                ensure!(self.last_batch_id < batch_id, "batch is not in the future");
```
Comment (fleupold) [inline driver/src/orderbook/streamed/state.rs:62]: with the strict inequality I'm a little worried about race conditions here, if we thought something was in a future, but happens to be `Current` now, I think that should be still fine.

**Review by fleupold — APPROVED** [review]
Nice 🍨

File: driver/src/orderbook/streamed/state.rs [hunk 3]
```diff
@@ -33,8 +45,31 @@ struct LastSolution {
 }
 
 impl State {
-    /// Can be used to create a `crate::models::AccountState`.
-    pub fn account_state(
+    /// Errors if State has only received a partial solution which would make the result
+    /// inconsistent.
+    /// Errors if the batch is `Future` but is not actually in the future.
+    /// Account balances that overflow a U256 are skipped.
+    pub fn orderbook(
```
Comment (fleupold) [inline driver/src/orderbook/streamed/state.rs:52]: uber naming nit: I think in the follow up PR it reads slightly nicer if this was called `get_orderbook` or even `get_orderbook_for_batch`.

File: driver/src/orderbook/streamed/state.rs (hunk 1 above)
Comment (nlordell) [reply driver/src/orderbook/streamed/state.rs:28]: This concerns me as it means there are times where the orderbook is unqueryable for future batches.

**Review by nlordell — APPROVED** [review]

File: driver/src/orderbook/streamed/state.rs (hunk 1 above)
Comment (nlordell) [reply driver/src/orderbook/streamed/state.rs:28]: Maybe the correct approach is to not include proceeds if a solution is pratially applied instead of erroring?

File: driver/src/orderbook/streamed/state.rs (hunk 3 above)
Comment (nlordell) [reply driver/src/orderbook/streamed/state.rs:52]: I like `orderbook_for_batch`. I'm fine with leaving out the `get_` is its pretty common in Rust for getters to simply be called what is being retrieved (such as `str::chars` instead of `str::get_chars`).

File: driver/src/orderbook/streamed/state.rs (hunk 1 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs:28]: I see your point. Since that is getting less straightforward and we are not actually going to solve optimistically yet I think we should restrict this to `Batch::Current` and revisit this later.

File: driver/src/orderbook/streamed/state.rs [hunk 4]
```diff
@@ -33,8 +45,31 @@ struct LastSolution {
...
+        &self,
+        batch: Batch,
+    ) -> Result<(
+        impl Iterator<Item = ((UserId, TokenId), U256)> + '_,
+        impl Iterator<Item = ModelOrder> + '_,
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs:57]: Is this a tuple of results or either or?

File: driver/src/orderbook/streamed/state.rs [hunk 5]
```diff
@@ -79,6 +113,8 @@ impl State {
     /// `block_batch_id` is the current batch based on the timestamp of the block that contains the
     ///  event.
     pub fn apply_event(mut self, event: &Event, block_batch_id: BatchId) -> Result<Self> {
+        ensure!(self.last_batch_id <= block_batch_id, "event in the past");
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs:118]: I see a lot of `ensure` happening... is this the same as assert? Are these directly relevant to this PR?

File: driver/src/orderbook/streamed/state.rs [hunk 6]
```diff
@@ -198,6 +234,8 @@ impl State {
         sell_balance_fn: impl FnOnce(&mut Balance) -> Result<()>,
         buy_balance_fn: impl FnOnce(&mut Balance) -> Result<()>,
     ) -> Result<()> {
+        self.solution_partially_received = true;
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs:239]: What is the meaning and purpose of a partially received solution? I see it is set true here and then false again later. Is this flag only accounting for the time between here and a few lines down the road? If this is the case, does this flag act as a sort-of lock for some sort of threaded environment?

File: driver/src/orderbook/streamed/state.rs (hunk 1 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs:28]: Hmm but we do need the batch id in case the it has increased but we just didn't get an event for the next batch yet. Then I would still say we don't fix the optimistic solving case yet because it is more complicated but keep the code as is until we need optimistic solving or the event based orderbook has been merged and we have more time.

File: driver/src/orderbook/streamed/state.rs [hunk 7]
```diff
@@ -652,4 +691,93 @@ mod tests {
             2
         );
     }
+
+    #[test]
+    fn orderbook_batch_id() {
+        let mut state = state_with_fee();
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs:699]: I was not aware of `state_with_fee` is this introduced here? I thought that constructors like this always took the form `State::new_with_fee`

File: driver/src/orderbook/streamed/state.rs [hunk 8]
```diff
@@ -652,4 +691,93 @@ mod tests {
...
+        let event = Deposit {
+            user: address(3),
+            token: address(0),
+            amount: 1.into(),
+            batch_id: 0,
+        };
+        state = state.apply_event(&Event::Deposit(event), 0).unwrap();
+        let balance = state.orderbook(Batch::Current).unwrap().0.next().unwrap().1;
+        assert_eq!(balance, U256::zero());
+        assert!(state.orderbook(Batch::Future(0)).is_err());
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs]: Is this an error because there is no way 0 is in the future?

File: driver/src/orderbook/streamed/state.rs [hunk 9]
```diff
@@ -652,4 +691,93 @@ mod tests {
...
+        let balance = state
+            .orderbook(Batch::Current)
+            .unwrap()
+            .0
+            .collect::<HashMap<_, _>>();
+        assert_eq!(balance.get(&(address(2), 0)), Some(&10.into()));
+        assert_eq!(balance.get(&(address(2), 1)), Some(&10.into()));
+
+        // Error because the solution is not complete.
+        assert!(state.orderbook(Batch::Future(2)).is_err());
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs]: Why is it not complete?

File: driver/src/orderbook/streamed/state.rs [hunk 10]
```diff
@@ -652,4 +691,93 @@ mod tests {
...
+            .apply_event(&Event::SolutionSubmission(event), 1)
+            .unwrap();
+
+        let balance = state
+            .orderbook(Batch::Future(2))
+            .unwrap()
+            .0
+            .collect::<HashMap<_, _>>();
+        assert_eq!(balance.get(&(address(2), 0)), Some(&12.into()));
+        assert_eq!(balance.get(&(address(2), 1)), Some(&9.into()));
```
Comment (bh2smith) [inline driver/src/orderbook/streamed/state.rs:789]: Nice!

**Review by bh2smith — APPROVED** [review]
Looks pretty great. Have a few inline questions and comments.

File: driver/src/orderbook/streamed/state.rs (hunk 4 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs:57]: Because the Result is on the outside it is a Result of a tuple. So either you have both elements of the tuple or an error.

File: driver/src/orderbook/streamed/state.rs (hunk 5 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs:118]: It is like an assert but intead of panicking it returns an error.

You can check where the ensures come from. It is at the top `use anyhow::{anyhow, bail, ensure, Result};`. The documentation for that is here https://docs.rs/anyhow/1.0.28/anyhow/macro.ensure.html . Or you can try `cargo doc --open` which generates the documentation locally and then searching for `ensure`.

File: driver/src/orderbook/streamed/state.rs (hunk 6 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs:239]: The intention is to notice if someone tries to get the orderbook from us while we have received only some trades of a solution but not all of them. In that case the orderbook wouldn't be correct.

File: driver/src/orderbook/streamed/state.rs (hunk 7 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs:699]: It is just a function that was already there from a previous test. If you open the whole file you can see it.

File: driver/src/orderbook/streamed/state.rs (hunk 8 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs]: It was an error (just changed it) because batch id 0 is not actually in the future and the code creating the orderbook checked that.

File: driver/src/orderbook/streamed/state.rs (hunk 9 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs]: Because we got Trade events but no SolutionSubmission event which means there could be more trades.

File: driver/src/orderbook/streamed/state.rs (hunk 1 above)
Comment (nlordell) [reply driver/src/orderbook/streamed/state.rs:28]: Do you mind opening an issue so we don't forget?

File: driver/src/orderbook/streamed/state.rs (hunk 9 above)
Comment (bh2smith) [reply driver/src/orderbook/streamed/state.rs]: Does the `SolutionSubmission` event contain any relevant information regarding how many trades we expect there to be?

**Comment by bh2smith** [comment]
Looks ready for merger!

File: driver/src/orderbook/streamed/state.rs (hunk 9 above)
Comment (e00E) [reply driver/src/orderbook/streamed/state.rs]: No

**Comment by fleupold** [comment]
> I notice that the tests feels needlessly long. In the future I want to refactor the tests into more helper functions and macros to make them smaller and easier to read.

Sounds good. Generally, I'd be careful of over-optimizing test code too much (in terms of trying to avoid duplicate code). It can quickly make the test very hard to understand. I generally like having all information that is relevant for a specific test inside the single function rather than having to jump through a bunch of common setup methods to see what the given state/expectations actually is/are. 
That being said, using some macros or the like to make them shorter is probably a cool idea.
